### DISCUSSION TITLE
Best practices for unit-testing a pointcloud component using pointcloud Layer

### DISCUSSION THREAD

**Discussion by iamravisingh** [visgl/deck.gl#9141] [Q&A] [answered]
Hello Everyone,

I'm currently working on a project where we're using the DeckGL `PointCloud` layer to render 3D point cloud data from .las files. I'm looking to write unit tests for this component, and I'm particularly interested in best practices for doing so effectively.

**Context**
Testing Framework: I'm using `Cypress` to test the components of our unit tests.
Mocking Data: Since we're dealing with .las files, my approach is to mock this data to simulate the point cloud rendering during tests.

**Questions**
1. Mocking .las Files: Has anyone successfully mocked .las files for testing? If so, what tools or strategies did you use?
2. What are the recommended ways to test user interactions with the point cloud (e.g., selecting points, hovering over them)? 
Should these be tested at the unit level or in higher-level integration tests?
4. Are there any performance concerns when running unit tests that simulate large point clouds? How can we mitigate these during testing?

**Comment by ibgreen** [comment, marked as answer]
Rather general question, don't really have an answer but maybe I can give you a few pointers...

> Mocking .las Files: Has anyone successfully mocked .las files for testing? If so, what tools or strategies did you use?

Especially in loaders.gl we have a lot of test files for different formats. We typically include smaller test files in the primary repo, an load them from the file system. You can typically `fetch` local resources from your tests, or you can use the "rawgithub" URLs to access your files as if GitHub was a CDN.

For bigger files, we have separate repos like deck.gl-data, to avoid making the code repo "too heavy".

> What are the recommended ways to test user interactions with the point cloud (e.g., selecting points, hovering over them)?

deck.gl does have some interaction tests, taking a look at those could be a starting point.

> Should these be tested at the unit level or in higher-level integration tests?

These tests can at least be written in the same way as unit tests, again see what deck does.

> Are there any performance concerns when running unit tests that simulate large point clouds? How can we mitigate these during testing?

You'll need to try and decide what is acceptable to you loading local files is pretty fast compared to loading from a network. A really big slow test could be kept as a separate test command and you'd probably not run that one in e.g. your CI testing for every commit.

**Reply by iamravisingh** [reply]
@ibgreen thanks for looking into the questions and for your thoughts as pointers. I'll check how deckgl is written on unit level.
